### ISSUE TITLE
frequent network connections, many over port 80?

### ISSUE THREAD

**Issue by bgentry** [keybase/keybase-issues#2782]
In the past few days, Keybase has been making frequent outbound network connections. I've tracked these using Little Snitch. It's as if it keeps re-verifying my contacts or something. Many of these connections occur via port 80 to verify custom domains.

I've been running Little Snitch alongside Keybase for a long time, and these only started happening a few days ago. I'd hoped that it might just be some kind of one-time re-verification, but I've seen it hitting the same domains multiple times now.

Is this expected? Will this go on forever as long as I have the Keybase app open?

**Comment by maxtaco** [comment]
We recently starting checking contacts in the background.  There is currently one check every 3 minutes.  If this is a problem, let us know, it's easy enough to disable it. We like it on by default because it makes people's experience more secure --- you get warned eagerly if someone's identity has broken due to broken proofs.

CC: @malgorithms

**Comment by bgentry** [comment]
I'm concerned about a few things related to this. One is the sheer amount of network chatter generated by this service checking so frequently. Another is the fact that many of these checks are going to custom domains over port 80, domains which can be altered by other people at any time. If I'm on a hostile network, I'd rather not leave such a big unencrypted and identifiable footprint.

Are you not concerned about the identifiable footprint left by a specific user's graph of contacts?

Another concern. Say I validate a custom domain that I'm running for free on an app platform like Heroku. When I add the domain, I don't expect that it will receive any regular traffic as a result of that. Maybe a couple requests whenever a new contact verifies me, which is fairly rare. But if all my contacts are regularly re-checking my site, that's enough to prevent my app from idling and either run up a bill or at least exhaust all of my free dyno hours on Heroku.

I would be a little less surprised if this was only regularly verifying contacts that I'm actually communicating with in some way. As it stands today, I'm not actively using Keybase for anything but I like keeping it around for when I will. This feature is enough for me to stop running Keybase unless I'm actively using it.

I'm sure you've probably thought through your desired security model more than I have. Is it not possible to have keybase's servers do periodic verification of all contacts in the system, and then gossip that info out to clients for full verification in the event of any failures?

Most of my concerns actually seem related to verifying arbitrary custom domains. Maybe the answer is to treat these differently than other verification methods like Twitter, HN, Github, etc?

**Comment by scoates** [comment]
My problem might be a separate issue (and I'm happy to open a new one if you keybase folks would like), but I found this one and it seems similar.

I use Little Snitch as well, and I have a low bandwidth profile that I use when I'm tethered with my phone. I use this profile to prevent Dropbox updates, etc. I currently have keybase blocked on this profile (to prevent it from using up my data transfer).

I'm currently tethering, and I see that `keybase` and `kbfs` are making *very* frequent (denied) requests to `gregor.kbfs.keybase.io` (`:443` and `:80`) and `mdserver.kbfs.keybase.io` (`:443`). This is happening at least once per second—not every 3 minutes.

I suspect something is retrying these connections very frequently if it thinks it is not connected. I suggest you add a cooldown period or otherwise make the failed-to-connect timeout much longer. Maybe once a minute. I have to quit the keybase menubar app (Mac) to get it to stop.

![keybase-little-snitch](https://cloud.githubusercontent.com/assets/71983/22397540/6267298a-e541-11e6-81bf-a97764a269ed.gif)
(this clip is realtime)

**Comment by maxtaco** [comment]
These might be network pings (at most 1k of data) to keep network connections alive.  Cc @mmaxim would know for sure.

**Comment by mmaxim** [comment]
If the connections to those servers are blocked, the clients will keep trying to reconnect fairly aggressively now. This is a recent change to make KBFS and chat more responsive, but we can talk about some ways we can find a middle ground to help address this particular case.

**Comment by bgentry** [comment]
Exponential backoff with a max interval of 30s or a couple minutes is a reasonable approach for that. Also useful to avoid a thundering herd if your service ever goes offline :) But yeah, separate issue.

**Comment by maxtaco** [comment]
We used to have that but unfortunately it performs poorly for the common
case which is intermittent disconnectedness.

**Comment by cjb** [comment]
Hi @bgentry!  Sorry we didn't reply here earlier.  Based on your feedback, we reverted this change that turned on background identifying, the revert update went out to users early last week.  We're going to think harder about this before considering turning it back on again.

**Comment by bgentry** [comment]
@cjb thanks for the update, I believe that's the right decision for a privacy-by-default design.

Sorry for spreading misinformation. I updated my HN comment to reflect the updated reality. Please keep up the good work :)

**Comment by bgentry** [comment]
If this is disabled, should I still be seeing connections like these from keybase?

<img width="735" alt="screen shot 2017-02-08 at 12 27 39 pm" src="https://cloud.githubusercontent.com/assets/114033/22755924/1c95a686-edfa-11e6-8b11-ae0e2bc2bf66.png">

I haven't used the app to actually do anything recently.

**Comment by cjb** [comment]
@bgentry I'd guess you might not have applied an update yet.  The first update check happens an hour after you start the app.

Once it's applied, you should see:
```
$ keybase version
Client:  1.0.18-20170208165022+dc66461
Service: 1.0.18-20170208165022+dc66461
```
Does that help?

**Comment by bgentry** [comment]
@cjb you're right, I'd assumed the update check would happen right away. Also explains why I'm still on a weeks old version, because each time I opened the app and saw it immediately use port 80, I closed it 😄 

I've now updated Keybase to the version you mentioned.

**Comment by jedahan** [comment]
I think this can be closed now?
